File: leetcode/162_find_peak_element/Solution.java
import java.util.Arrays;

class Solution {
    public int findPeakElement(int[] nums) {
        // linear
        // int prev = -999999999;
        // for (int i = 0; i < nums.length; i++) {
        //     if ((i-1 >= 0 && nums[i] > nums[i-1]) || i-1 < 0) {
        //         if ((i+1 <= nums.length-1 && nums[i] > nums[i+1]) || i+1 >= nums.length) {
        //             System.out.println(i);
        //             return i;
        //         }
        //     }
        // }
        // return -1;

        // logarithmic
        int floor = 0;
        int ceiling = nums.length - 1;
        int midIdx = Math.floorDiv(ceiling + floor, 2);

        while (true) {
            boolean leftIsLower = true;
            boolean rightIsLower = true;

            if (midIdx != 0 && nums[midIdx] <= nums[midIdx-1]) leftIsLower = false;
            if (midIdx != nums.length - 1 && nums[midIdx] <= nums[midIdx+1]) rightIsLower = false;

            if (leftIsLower && rightIsLower) {
                // System.out.println(midIdx);
                return midIdx;
            }
            
            if (!leftIsLower) { // left is higher, go left
                ceiling = midIdx;
                if (midIdx - floor <= 1) midIdx = floor;
                else midIdx = Math.floorDiv(midIdx + floor, 2);
            } else { // right is higher, go right
                floor = midIdx;
                if (ceiling - midIdx <= 1) midIdx = ceiling;
                else midIdx = Math.floorDiv(midIdx + ceiling, 2);
            }
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        assert sol.findPeakElement(new int[]{1,2,3,1}) == 2;
        assert Arrays.asList(1, 5).contains(sol.findPeakElement(new int[]{1,2,1,3,5,6,4}));
        assert sol.findPeakElement(new int[]{-123123123}) == 0;
        assert sol.findPeakElement(new int[] {1,2,3,4,5,6,7,8,9,10,11,12}) == 11;
        assert sol.findPeakElement(new int[]{13,12,11,10,9,8,7,6,5,4,3,2,1,0}) == 0;
    }
}

File: leetcode/162_find_peak_element/solution.py
from typing import List

class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        # linear
        # prev = -9999999999
        # for idx, num in enumerate(nums):
        #     if (idx > 0 and num > prev) or idx == 0:
        #         if idx >= len(nums) - 1 or num > nums[idx+1]:
        #             return idx

        #     prev = num

        # logarithmic
        if len(nums) == 1:
            return 0
        
        mid_idx = len(nums) // 2
        floor = 0
        ceiling = len(nums) - 1

        while True:
            left_is_lower, right_is_lower = True, True
            mid_num = nums[mid_idx]
            if mid_idx - 1 >= 0 and nums[mid_idx - 1] >= mid_num:
                left_is_lower = False
            if mid_idx + 1 <= len(nums) - 1 and mid_num <= nums[mid_idx + 1]:
                right_is_lower = False

            # print(f'mid_idx: {mid_idx}, mid_num: {mid_num}, left_is_lower: {left_is_lower}, right_is_lower: {right_is_lower}, floor: {floor}, ceiling: {ceiling}')

            if left_is_lower and right_is_lower:
                return mid_idx
            else:
                if not left_is_lower: # left is higher, go left
                    ceiling = mid_idx
                    if mid_idx - floor <= 1:
                        mid_idx = floor
                    else:
                        mid_idx = (mid_idx + floor) // 2
                else: # right is higher, go right
                    floor = mid_idx
                    if ceiling - mid_idx <= 1:
                        mid_idx = ceiling
                    else:
                        mid_idx = (mid_idx + ceiling) // 2


if __name__ == '__main__':
    sol = Solution()
    # print(sol.findPeakElement([13,12,11,10,9,8,7,6,5,4,3,2,1,0]))
    assert sol.findPeakElement([1,2,3,1]) == 2
    assert sol.findPeakElement([1,2,1,3,5,6,4]) in (1, 5)
    assert sol.findPeakElement([-123123123]) == 0
    assert sol.findPeakElement([1,2,3,4,5,6,7,8,9,10,11,12]) == 11
    assert sol.findPeakElement([13,12,11,10,9,8,7,6,5,4,3,2,1,0]) == 0
